src/pyodide-py/pyodide/webloop.py

```diff
@@ -0,0 +1,425 @@
+        * support contextvars
+        * add event loop policy
+        * throw error when the loop is already running
+        * support Python 3.8
+        * fix run_until_complete
+"""
+
+import heapq
+import asyncio
+import time
+import traceback
+import contextvars
+
+import js
+from typing import Dict, Tuple, Optional, Awaitable, Callable
+
+
+class WebLoop(asyncio.AbstractEventLoop):
+    """A custom event loop for running asyncio in Pyodide
+
+    It works by utilizing the browser event loop via the setTimeout function
+    """
+
+    def __init__(self, debug: Optional[bool] = False, interval: Optional[int] = 10):
+        """
+        Instatiate the web loop
+        """
+        self._running = False
+        self._immediate = []
+        self._scheduled = []
+        self._futures = []
+        self._next_handle = None
+        self._debug = debug
+        self._stop = False
+        self._interval = interval
+        self._timeout_promise = js.eval(
+            "self._timeoutPromise = function(time){return new Promise((resolve)=>{setTimeout(resolve, time);});}"
+        )
+        self._exception_handler = None
+        self._task_factory = self._default_task_factory
+        self._stop_callbacks = None
+        self._result = None
+        self._exception = None
+
+    def get_debug(self):
+        """
+        Get the debug mode (bool) of the event loop.
+        """
+        return self._debug
+
+    def set_debug(self, enabled: bool):
+        """
+        Set the debug mode of the event loop.
+        """
+        self._debug = enabled
+
+    def time(self):
+        """
+        Return the current time, as a float value, according to the time module (time.time()).
+        """
+        return time.time()
+
+    def run_forever(self):
+        """
+        Run the event loop until stop() is called.
+
+        Note that this function is different from the standard asyncio loop implementation in two ways:
+         1) It won't block the execution
+         2) It returns a Promise object
+
+        """
+        if self._running:
+            raise RuntimeError("This event loop is already running")
+
+        def run(resolve, reject):
+            self._stop = False
+            if asyncio.get_event_loop() == self:
+                asyncio._set_running_loop(self)
+            self._stop_callbacks = (resolve, reject)
+            self._do_tasks(forever=True)
+
+        return js.Promise.new(run)
+
+    def run_until_complete(self, future: Awaitable):
+        """
+        Run until the future (an instance of Future) has completed.
+
+        Note that this function is different from the standard asyncio loop implementation in two ways:
+         1) It won't block the execution
+         2) It returns a Promise object
+
+        Parameters
+        ----------
+        future
+        A future or coroutine object
+
+
+        Returns
+        -------
+        A Promise object with the returned result or exception
+        """
```

Comment:
A simpler implementation could be to wrap the future in a promise, so get the browser event loop to schedule it.
My PR #880 adds conversion promise ==> future. We should add conversion future ==> promise. This will again defer work to the browser event loop.

Reply:
Interesting idea, but since run_until_complete is linked with other functions such as `create_task` and `call_soon` (e.g. in the coroutine one can call create task and we need wait also for the newly created tasks), it is likely we need to fix the promise cancellation issue.

Reply:
Let's continue the discussion below.